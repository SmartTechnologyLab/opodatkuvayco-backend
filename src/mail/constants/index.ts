export const mailConstants = {
  host: process.env.SMTP_HOST,
  user: process.env.SMTP_USER,
  pass: process.env.SMTP_EMAIL_PASSWORD,
};
