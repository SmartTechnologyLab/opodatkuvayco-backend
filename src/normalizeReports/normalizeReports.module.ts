import { Module } from '@nestjs/common';
import { NormalizeReportsService } from './normalizeReports.service';

@Module({
  providers: [NormalizeReportsService],
})
export class NormalizeReportsModule {}
