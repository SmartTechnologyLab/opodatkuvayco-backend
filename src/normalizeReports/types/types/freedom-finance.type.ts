export type FreedomFinanceCorporateActionTypeId = 'dividend' | 'maturity';
