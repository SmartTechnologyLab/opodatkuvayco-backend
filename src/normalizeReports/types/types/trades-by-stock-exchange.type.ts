// import { StockExchangeEnum } from 'src/normalizeTrades/constants/enums';

// export type ReportByStockExchange = {
//   [StockExchangeEnum.FREEDOM_FINANCE]: IFreedomFinanceReport[];
//   [StockExchangeEnum.IBRK]: IbkrReport[];
// };
