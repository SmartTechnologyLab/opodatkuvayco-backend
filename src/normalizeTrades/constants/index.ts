export enum StockExchange {
  FREEDOM_FINANCE = 'FreedomFinance',
}
