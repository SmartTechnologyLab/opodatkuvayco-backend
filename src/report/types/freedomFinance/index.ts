import { OperationType } from '..';

type FreedomFinanceCorporateActionTypeId = 'dividend' | 'maturity';

export interface IFreedomFinanceCorporateAction {
  date: string;
  type: string;
  type_id: FreedomFinanceCorporateActionTypeId;
  amount: number;
  ticker: string;
  isin: string;
  currency: string;
}

export interface IFreedomFinanceTrade {
  date: string;
  p: number;
  q: number;
  commission: number;
  operation: OperationType;
  instr_nm: string;
  curr_c: string;
}

export interface IFreedomFinanceReport {
  date_start: string;
  corporate_actions: {
    detailed: IFreedomFinanceCorporateAction[];
  };
  trades: {
    detailed: IFreedomFinanceTrade[];
  };
}
