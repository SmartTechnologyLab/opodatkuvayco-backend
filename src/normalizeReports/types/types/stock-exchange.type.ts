import { StockExchangeEnum } from 'src/normalizeTrades/constants/enums';

export type StockExchangeType =
  | StockExchangeEnum.FREEDOM_FINANCE
  | StockExchangeEnum.IBRK;
