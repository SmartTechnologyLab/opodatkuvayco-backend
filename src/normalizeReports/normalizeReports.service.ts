import { Injectable } from '@nestjs/common';
import { StockExchange } from '../normalizeTrades/constants';
import { IReport } from '../report/types';
import {
  IFreedomFinanceReport,
  IFreedomFinanceTrade,
} from '../report/types/freedomFinance';

@Injectable()
export class NormalizeReportsService {
  constructor() {}

  getReportByStockExchange(
    report: unknown,
    stockExchange: StockExchange,
  ): IReport<unknown> {
    switch (stockExchange) {
      case StockExchange.FREEDOM_FINANCE:
        return this.normalizeFreedomFinanceReport(
          report as IFreedomFinanceReport,
        );
      default:
        break;
    }
  }

  normalizeFreedomFinanceReport(
    report: IFreedomFinanceReport,
  ): IReport<IFreedomFinanceTrade> {
    return {
      dateStart: report.date_start,
      trades: report.trades.detailed,
    };
  }
}
