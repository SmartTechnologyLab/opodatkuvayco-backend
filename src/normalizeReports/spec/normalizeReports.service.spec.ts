import { IFreedomFinanceReport } from 'src/report/types/freedomFinance';
import { NormalizeReportsService } from '../normalizeReports.service';
import { StockExchange } from '../../normalizeTrades/constants';

describe('Normalize Report Service', () => {
  let normalizeReportsService: NormalizeReportsService;

  beforeEach(() => {
    normalizeReportsService = new NormalizeReportsService();
  });

  describe('getReportByStockExchange should return normalize report', () => {
    it('when stockExchange is freedom finance', () => {
      const freedomFinanceReport: IFreedomFinanceReport = {
        date_start: '2021-01-01',
        trades: {
          detailed: [
            {
              date: '2021-01-01',
              instr_nm: 'AAPL',
              operation: 'buy',
              p: 100,
              q: 1,
              curr_c: 'USD',
              commission: 0,
            },
          ],
        },
        corporate_actions: {
          detailed: [],
        },
      };

      const report = normalizeReportsService.getReportByStockExchange(
        freedomFinanceReport,
        StockExchange.FREEDOM_FINANCE,
      );

      expect(report).toMatchSnapshot();
    });
  });
});
