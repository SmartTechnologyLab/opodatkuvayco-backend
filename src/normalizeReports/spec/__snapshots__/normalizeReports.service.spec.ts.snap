// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Normalize Report Service getReportByStockExchange should return normalize report when stockExchange is freedom finance 1`] = `
{
  "dateStart": "2021-01-01",
  "trades": [
    {
      "commission": 0,
      "curr_c": "USD",
      "date": "2021-01-01",
      "instr_nm": "AAPL",
      "operation": "buy",
      "p": 100,
      "q": 1,
    },
  ],
}
`;
